### PR TITLE
[SYCL] Implement SYCL part of sycl_ext_oneapi_prefetch

### PR DESCRIPTION
**Pull request by KornevNikita**
Spec: https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_prefetch.asciidoc

The joint_prefetch functions will be done is a separate patch.

This implementation also requires a patch for llvm-spirv translator. SPIRVWriter should handle these new annotations and create the appropriate decorations in spv representation.

### CODE REVIEW

File: llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp [hunk 1]
```diff
@@ -619,6 +619,12 @@ bool CompileTimePropertiesPass::transformSYCLPropertiesAnnotation(
     if (*Property.first == "sycl-alignment")
       continue;
 
+    // leave these annotations as is. They will be processed by SPIRVWriter.
```
Comment (aelovikov-intel) [inline llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp:622]:
```suggestion
    // Leave these annotations as is. They will be processed by SPIRVWriter.
```

File: llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp [hunk 2]
```diff
@@ -619,6 +619,12 @@ bool CompileTimePropertiesPass::transformSYCLPropertiesAnnotation(
...
+    if (*Property.first == "sycl-prefetch-hint" ||
```
Comment (aelovikov-intel) [inline llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp:623]: I assume the type of `Property` is an `std::pair`. If so, please change line 616 to use structured bindings. Can be done either here or in another PR merged before this one.

File: sycl/include/sycl/ext/oneapi/prefetch.hpp [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+inline constexpr prefetch_hint_key::value_t<cache_level::L2, nontemporal>
+    prefetch_hint_L2_nt;
+inline constexpr prefetch_hint_key::value_t<cache_level::L3, nontemporal>
+    prefetch_hint_L3_nt;
+inline constexpr prefetch_hint_key::value_t<cache_level::L4, nontemporal>
+    prefetch_hint_L4_nt;
+
+template <typename T, cache_level Level, typename Hint>
+struct is_valid_property<T, prefetch_hint_key::value_t<Level, Hint>>
+    : std::bool_constant<std::is_pointer<T>::value> {};
```
Comment (aelovikov-intel) [inline sycl/include/sycl/ext/oneapi/prefetch.hpp:57]: Why `std::is_pointer` and not a SYCL type trait that would accept `multi_ptr` as well? Please put that into a comment in the implementation.

File: sycl/include/sycl/ext/oneapi/prefetch.hpp [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+namespace detail {
+template <> struct IsCompileTimeProperty<prefetch_hint_key> : std::true_type {};
+
+template <cache_level Level, typename Hint>
+struct PropertyMetaInfo<prefetch_hint_key::value_t<Level, Hint>> {
+  static constexpr const char *name = std::is_same_v<Hint, nontemporal>
+                                          ? "sycl-prefetch-hint-nt"
+                                          : "sycl-prefetch-hint";
+  static constexpr int value = static_cast<int>(Level);
```
Comment (aelovikov-intel) [inline sycl/include/sycl/ext/oneapi/prefetch.hpp:67]: I'd like to see explicit values in the `cache_level` enum definition because of this usage.

File: sycl/include/sycl/ext/oneapi/prefetch.hpp [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+};
+
+template <typename Properties>
+void prefetch_impl(void *ptr, size_t bytes, Properties properties) {
+#ifdef __SYCL_DEVICE_ONLY__
```
Comment (aelovikov-intel) [inline sycl/include/sycl/ext/oneapi/prefetch.hpp:72]: I think this will be causing "unused parameter" warnings in the host compilation.

File: sycl/include/sycl/ext/oneapi/prefetch.hpp [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+  prefetch((void *)ptr, count * sizeof(T), properties);
+}
+
+// Only available if AddressSpace == global_space || AddressSpace ==
+// generic_space
+template <
+    access::address_space AddressSpace, access::decorated IsDecorated,
+    typename Properties = empty_properties_t,
+    std::enable_if_t<AddressSpace == access::address_space::global_space ||
+                     AddressSpace == access::address_space::generic_space>>
```
Comment (aelovikov-intel) [inline sycl/include/sycl/ext/oneapi/prefetch.hpp:115]: I'd move that into the return type. Also, a helper would be useful too.

File: sycl/include/sycl/ext/oneapi/properties/properties.hpp [hunk 7]
```diff
@@ -187,6 +187,8 @@ properties(PropertyValueTs... props)
     -> properties<typename detail::Sorted<PropertyValueTs...>::type>;
 #endif
 
+using empty_properties_t = decltype(properties{});
```
Comment (aelovikov-intel) [inline sycl/include/sycl/ext/oneapi/properties/properties.hpp:190]: I'd prefer this to go in a separate PR updating existing places where it can be used.

File: sycl/test/extensions/prefetch.cpp [hunk 8]
```diff
@@ -0,0 +1,38 @@
+// RUN: %clangxx -fsycl-device-only -S %s -o - | FileCheck %s
+
+#include <sycl/sycl.hpp>
+
+char data[] = {0, 1, 2, 3};
+
+// CHECK: [[PREFETCH_STR:@.*]] = private unnamed_addr addrspace(1) constant [19 x i8] c"sycl-prefetch-hint\00", section "llvm.metadata"
```
Comment (aelovikov-intel) [inline sycl/test/extensions/prefetch.cpp:7]: Are these manually written? If so, do we have an utility script to generate them automatically? If yes, one might need to disable instrumentation/use `-O1` to get a more readable IR.

File: sycl/test/extensions/prefetch.cpp [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+// CHECK: [[ANNOTATION2:@.*]] = private unnamed_addr addrspace(1) constant { ptr addrspace(1), ptr addrspace(1) } { ptr addrspace(1) [[PREFETCH_STR]], ptr addrspace(1) [[PREFETCH_LVL1]] }, section "llvm.metadata"
+// CHECK: [[PREFETCH_STR_NT:@.*]] = private unnamed_addr addrspace(1) constant [22 x i8] c"sycl-prefetch-hint-nt\00", section "llvm.metadata"
+// CHECK: [[PREFETCH_LVL2:@.*]] = private unnamed_addr addrspace(1) constant [2 x i8] c"2\00", section "llvm.metadata"
+// CHECK: [[ANNOTATION3:@.*]] = private unnamed_addr addrspace(1) constant { ptr addrspace(1), ptr addrspace(1) } { ptr addrspace(1) [[PREFETCH_STR_NT]], ptr addrspace(1) [[PREFETCH_LVL2]] }, section "llvm.metadata"
+
+int main() {
+  namespace syclex = sycl::ext::oneapi::experimental;
+  sycl::queue q;
+  void *dataPtr = &data;
+  q.parallel_for(1, [=](sycl::id<1> idx) {
```
Comment (aelovikov-intel) [inline sycl/test/extensions/prefetch.cpp:20]: I think this can be `single_task`. I'd also want to see an E2E test with this used in non-uniform control flow (I don't think the spec prohibits that).

File: llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp (hunk 1 above)
Comment (KornevNikita) [reply llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp]: [19e19a8](https://github.com/intel/llvm/pull/11458/commits/19e19a8d79a821da89a90a3a127bfdd1a5a975fd)

File: llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp (hunk 2 above)
Comment (KornevNikita) [reply llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp]: Is it what you assume to see [30e531a](https://github.com/intel/llvm/pull/11458/commits/30e531a17d6d59e337a85a36c22276be7d81ca10)? Or should we give some another names to _first_ and _second_ variables.

File: sycl/include/sycl/ext/oneapi/prefetch.hpp (hunk 3 above)
Comment (KornevNikita) [reply sycl/include/sycl/ext/oneapi/prefetch.hpp]: This trait is redundant now as I refused to use the annotated_ptr. Removed in [42de79f](https://github.com/intel/llvm/pull/11458/commits/42de79f1183b750d905557b77f53011bd7669abe)

File: sycl/include/sycl/ext/oneapi/prefetch.hpp (hunk 4 above)
Comment (KornevNikita) [reply sycl/include/sycl/ext/oneapi/prefetch.hpp:63]: [2d5a360](https://github.com/intel/llvm/pull/11458/commits/2d5a3607fdaab946b69b632dab85e3f085a29477)

File: sycl/include/sycl/ext/oneapi/prefetch.hpp (hunk 6 above)
Comment (KornevNikita) [reply sycl/include/sycl/ext/oneapi/prefetch.hpp]: [c40e4bd](https://github.com/intel/llvm/pull/11458/commits/c40e4bd3cba3b809272518fbd0922d475900ad92)

File: sycl/include/sycl/ext/oneapi/properties/properties.hpp (hunk 7 above)
Comment (KornevNikita) [reply sycl/include/sycl/ext/oneapi/properties/properties.hpp:190]: Oh, yes, I forgot to delete from here. I'll push a separate PR soon.

File: sycl/include/sycl/ext/oneapi/prefetch.hpp (hunk 5 above)
Comment (KornevNikita) [reply sycl/include/sycl/ext/oneapi/prefetch.hpp:68]: [c40e4bd](https://github.com/intel/llvm/pull/11458/commits/c40e4bd3cba3b809272518fbd0922d475900ad92)

File: sycl/test/extensions/prefetch.cpp (hunk 9 above)
Comment (KornevNikita) [reply sycl/test/extensions/prefetch.cpp]: Moved to single task - [7afe1c9](https://github.com/intel/llvm/pull/11458/commits/7afe1c961241cce2b09d4fc2ac2054f0accf3ccd). I'm going to add E2E tests a bit later when the whole feature (incl. llvm-spirv translator part) is done. Not sure what do you mean by "non-uniform control". Could you please explain a bit?

File: llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp (hunk 2 above)
Comment (aelovikov-intel) [reply llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp]: Almost... Now need to change "first" and "second" into something meaningful.

File: sycl/test/extensions/prefetch.cpp (hunk 9 above)
Comment (aelovikov-intel) [reply sycl/test/extensions/prefetch.cpp]:
```
if (id.get_global_id(0) % 3 == 0))
  syclex::prefetch(p);
```

File: llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp [hunk 10]
```diff
@@ -613,13 +613,18 @@ bool CompileTimePropertiesPass::transformSYCLPropertiesAnnotation(
...
     // sycl-alignment is converted to align on
     // previous parseAlignmentAndApply(), dropping here
-    if (*Property.first == "sycl-alignment")
+    if (first == "sycl-alignment")
       continue;
 
-    auto DecorIt = SpirvDecorMap.find(*Property.first);
+    // Leave these annotations as is. They will be processed by SPIRVWriter.
+    if (first == "sycl-prefetch-hint" || first == "sycl-prefetch-hint-nt") {
+      return false;
```
Comment (aelovikov-intel) [inline llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp:624]: This doesn't feel right to me, but it's outside of SYCL RT...

File: sycl/include/sycl/ext/oneapi/prefetch.hpp [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+  prefetch((void *)ptr, count * sizeof(T), properties);
+}
+
+// Only available if AddressSpace == global_space || AddressSpace ==
+// generic_space
+template <access::address_space AddressSpace, access::decorated IsDecorated,
+          typename Properties = empty_properties_t>
+std::enable_if_t<AddressSpace == access::address_space::global_space ||
+                     AddressSpace == access::address_space::generic_space,
+                 void>
```
Comment (aelovikov-intel) [inline sycl/include/sycl/ext/oneapi/prefetch.hpp:109]: I'd move that to a helper still. Maybe something like
```
template <address_space AS, bool condition = True>
detail::enable_prefetch_t;

detail::enable_prefetch_t<AS> overload1();
detail::enable_prefetch_t<AS, Dimensions >0> overload2();
```

File: sycl/include/sycl/ext/oneapi/prefetch.hpp [hunk 12]
```diff
@@ -0,0 +1,301 @@
...
+  prefetch((void *)ptr, sizeof(T), properties);
+}
+
+template <typename T, typename Properties = empty_properties_t>
+void prefetch(T *ptr, size_t count, Properties properties = {}) {
+  prefetch((void *)ptr, count * sizeof(T), properties);
+}
+
+// Only available if AddressSpace == global_space || AddressSpace ==
+// generic_space
```
Comment (aelovikov-intel) [inline sycl/include/sycl/ext/oneapi/prefetch.hpp:104]: This comment doesn't provide any additional information, please drop. Same below.

File: sycl/include/sycl/ext/oneapi/properties/properties.hpp (hunk 7 above)
Comment (KornevNikita) [reply sycl/include/sycl/ext/oneapi/properties/properties.hpp:190]: https://github.com/intel/llvm/pull/11476

File: sycl/include/sycl/ext/oneapi/prefetch.hpp (hunk 12 above)
Comment (KornevNikita) [reply sycl/include/sycl/ext/oneapi/prefetch.hpp]: [154660a](https://github.com/intel/llvm/pull/11458/commits/154660a893d44e868104238157f67e5aa5edd23d)

File: llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp (hunk 10 above)
Comment (KornevNikita) [reply llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp:640]: Looks like currently it's the only way to transform these properties into spirv decorations for this pointer. I tried to create `spirv.Decorations` metadata instead, but the compiler eliminates them with optimization flags.
Annotation also may be eliminated, but it's much less likely.

File: sycl/include/sycl/ext/oneapi/prefetch.hpp [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+                       void>;
+};
+
+template <access_mode AccessMode, bool AdditionalCondition>
+struct prefetch_helper<access_mode, AccessMode, AdditionalCondition> {
+  using type = std::enable_if_t<(AccessMode == access_mode::read ||
+                                 AccessMode == access_mode::write) &&
+                                    AdditionalCondition,
+                                void>;
+};
```
Comment (aelovikov-intel) [inline sycl/include/sycl/ext/oneapi/prefetch.hpp:80]: I didn't notice this part when I made my suggestion. I wouldn't want to use the same name (`prefetch_helper`) for both, I think we should two helpers with more explicit names.

Also, you can do

```
template <access::address_space AddressSpace, bool AdditionalCondition>
using prefetch_check_as_t = std::enable_if_t<...>;
```
Why don't we allow read-write mode?

File: sycl/include/sycl/ext/oneapi/prefetch.hpp [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+
+template <typename DataT, int Dimensions, access_mode AccessMode,
+          access::placeholder IsPlaceholder,
+          typename Properties = empty_properties_t>
+typename detail::prefetch_helper<access_mode, AccessMode,
+                                 (Dimensions > 0)>::type
+prefetch(
+    accessor<DataT, Dimensions, AccessMode, target::device, IsPlaceholder> acc,
+    id<Dimensions> offset, Properties properties = {}) {
+  prefetch((void *)&acc[offset], sizeof(DataT), properties);
```
Comment (aelovikov-intel) [inline sycl/include/sycl/ext/oneapi/prefetch.hpp:167]: Please use one of C++ casts.

File: sycl/include/sycl/ext/oneapi/prefetch.hpp [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+
+template <typename DataT, int Dimensions, access_mode AccessMode,
+          access::placeholder IsPlaceholder,
+          typename Properties = empty_properties_t>
+typename detail::prefetch_helper<access_mode, AccessMode,
+                                 (Dimensions > 0)>::type
+prefetch(
+    accessor<DataT, Dimensions, AccessMode, target::device, IsPlaceholder> acc,
+    size_t offset, size_t count, Properties properties = {}) {
+  prefetch((void *)&acc[offset], count * sizeof(DataT), properties);
```
Comment (aelovikov-intel) [inline sycl/include/sycl/ext/oneapi/prefetch.hpp:178]: Do we know for sure that `count` elements are consecutive in memory?

Comment (Pennycook) [reply sycl/include/sycl/ext/oneapi/prefetch.hpp:178]: We struggled a bit with multi-dimensional prefetches (see the [issues](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_prefetch.asciidoc#issues)).

Where we landed (for now) is that the block being prefetched is assumed to be contiguous, and only the offset itself is multi-dimensional.  It's effectively a shorthand to avoid computing the linear offset from the start of the buffer.  Note that the specification says for the multi-dimensional cases:

> Effects: Equivalent to prefetch((void*) &acc[offset], sizeof(DataT), properties).
> Effects: Equivalent to prefetch((void*) &acc[offset], count, properties).

If somebody requests a multi-dimensional prefetch later, we can describe it with a `range` parameter in place of a `size_t` count, and implement it the way you suggested (by calling the instruction multiple times).

File: sycl/include/sycl/ext/oneapi/prefetch.hpp (hunk 14 above)
Comment (KornevNikita) [reply sycl/include/sycl/ext/oneapi/prefetch.hpp]: Just copied it from the spec, but I guess we don't need them at all - [a3dbe96](https://github.com/intel/llvm/pull/11458/commits/a3dbe9608822613874122ca472efdd13884bbe57)

File: sycl/include/sycl/ext/oneapi/prefetch.hpp (hunk 13 above)
Comment (KornevNikita) [reply sycl/include/sycl/ext/oneapi/prefetch.hpp]: [Change helper](https://github.com/intel/llvm/pull/11458/commits/9e81d89e11539826a51efffa0bce91811cda2e47)

File: sycl/include/sycl/ext/oneapi/prefetch.hpp [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+}
+
+template <typename T, typename Properties = empty_properties_t>
+void prefetch(T *ptr, size_t count, Properties properties = {}) {
+  prefetch(ptr, count * sizeof(T), properties);
+}
+
+template <access::address_space AddressSpace, access::decorated IsDecorated,
+          typename Properties = empty_properties_t>
+std::enable_if_t<detail::check_prefetch_AS<AddressSpace>, void>
```
Comment (aelovikov-intel) [inline sycl/include/sycl/ext/oneapi/prefetch.hpp:115]:
```suggestion
std::enable_if_t<detail::check_prefetch_AS<AddressSpace>>
```

IIRC, `void` is the default. Same elsewhere.

**Review by aelovikov-intel — COMMENTED** [review]
Almost good. Please drop the unnecessary `void`s and the other concern is non-uniform control flow. I think I'll approve without the latter and we can follow-up on it separately.

File: sycl/include/sycl/ext/oneapi/prefetch.hpp (hunk 16 above)
Comment (KornevNikita) [reply sycl/include/sycl/ext/oneapi/prefetch.hpp]: 3e554ac

**Review by aelovikov-intel — APPROVED** [review]

**Comment by bader** [comment]
@intel/dpcpp-tools-reviewers, please, take a look. The compiler change is non-functional, so should be a no brainer.

File: sycl/include/sycl/ext/oneapi/prefetch.hpp
Comment (bader) [inline sycl/include/sycl/ext/oneapi/prefetch.hpp:1]: sycl/include/sycl/ext/oneapi/prefetch.hpp -> sycl/include/sycl/ext/oneapi/experimental/prefetch.hpp

File: llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp [hunk 17]
```diff
@@ -628,13 +628,13 @@ bool CompileTimePropertiesPass::transformSYCLPropertiesAnnotation(
   // Read the annotation values and create the new annotation string.
   std::string NewAnnotString = "";
   auto Properties = parseSYCLPropertiesString(M, IntrInst);
-  for (auto &Property : Properties) {
+  for (const auto &[propName, propVal] : Properties) {
```
Comment (sarnex) [inline llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp:631]: Can we use capital initial letters `PropName` `PropVal`, that's more standard in the codebase

**Review by asudarsa — COMMENTED** [review]
SYCLLowerIR changes look good to me, as they are NFC. Just a quick request about variable name. Also, I just noticed there is no test added.

Thanks

File: llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp (hunk 17 above)
Comment (KornevNikita) [reply llvm/lib/SYCLLowerIR/CompileTimePropertiesPass.cpp]: [c796b9e](https://github.com/intel/llvm/pull/11458/commits/c796b9e063313c04dcda28b0feeb46f08b0fbcff)

File: sycl/include/sycl/ext/oneapi/prefetch.hpp
Comment (KornevNikita) [reply sycl/include/sycl/ext/oneapi/prefetch.hpp:1]: [bf3d6c9](https://github.com/intel/llvm/pull/11458/commits/bf3d6c937f643cd7bb59f3e7b4d4045b70b7c8ed)

**Comment by AlexeySachkov** [comment]
Pre-commit failure doesn't seem related, it is covered by #11549
